union_pooling/union_pooling/union_pooler.py

```diff
@@ -0,0 +1,247 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2015, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import numpy
+
+from nupic.research.spatial_pooler import SpatialPooler
+
+
+
+REAL_DTYPE = numpy.float32
+
+
+
+class UnionPooler(SpatialPooler):
+  """
+  Experimental Union Pooler Python implementation. The Union Pooler builds a
+  "union SDR" of the most recent sets of active columns. It is driven by
+  active-cell input and, more strongly, by predictive-active cell input. The
+  latter is more likely to produce active columns. Such winning columns will
+  also tend to persist longer in the union SDR.
+  """
+
+
+  def __init__(self,
+               inputDimensions=[32,32],
+               columnDimensions=[64,64],
+               potentialRadius=16,
+               potentialPct=0.9,
+               globalInhibition=True,
+               localAreaDensity=-1.0,
+               numActiveColumnsPerInhArea=20.0,
+               stimulusThreshold=2,
+               synPermInactiveDec=0.01,
+               synPermActiveInc=0.03,
+               synPermConnected=0.3,
+               minPctOverlapDutyCycle=0.001,
+               minPctActiveDutyCycle=0.001,
+               dutyCyclePeriod=1000,
+               maxBoost=1.0,
+               seed=42,
+               spVerbosity=0,
+               wrapAround=True,
+
+               # union_pooler.py parameters
+               activeOverlapWeight=1.0,
+               predictedActiveOverlapWeight=10.0,
+               maxUnionActivity=0.20,
```

Comment:
Here's why I don't think so: this parameter serves as an upper bound on the ratio of active cells to total cells. So the min sparsity would vary e.g., from 0.0 to 0.2, so this name would suggest that min sparsity can be controlled. I also thought of using 'sparsity' in the name, but decided against it since I don't consider 20% sparse -- do you?